### PR TITLE
[Feature] Support ViTDet in projects

### PR DESCRIPTION
**Pull request by zwhus**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: projects/ViTDet/REMEAD.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+Based on mmdetection, this project aligns the test accuracy of the [ViTDet](https://github.com/facebookresearch/detectron2/tree/main/projects/ViTDet).
+<br>
+If you want to reproduce the test results, you need to convert model weights first, then run the test command.
+<br>
+The training accuracy will also be aligned with the official in the future
+
+|                     Method                      | Backbone | Pretrained Model |  Training set  |   Test set   | Epoch | Val Box AP | Val Mask  AP | Official Box AP | Official Mask AP | Download |
+| :---------------------------------------------: | :------: | :--------------: | :------------: | :----------: | :---: | :--------: | :----------: | :-------------: | :--------------: | :------: |
+| [ViTDet](./configs/mask-rcnn_vitdet_b_100ep.py) |  ViT-B   |       MAE        | COCO2017 Train | COCO2017 Val |  100  |    51.4    |     45.8     |      51.6       |       45.9       | model()  |
```
Comment (ZwwWayne) [inline projects/ViTDet/REMEAD.md:35]: mask-rcnn_vitdet_b_100ep.p -> vitdet_mask-rcnn_vit-b-mae_lsj-100e.py

File: projects/ViTDet/configs/lsj-100e_coco-instance.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+#     backend='petrel',
+#     path_mapping=dict({
+#         './data/': 's3://openmmlab/datasets/detection/',
+#         'data/': 's3://openmmlab/datasets/detection/'
+#     }))
+
+train_pipeline = [
+    dict(type='LoadImageFromFile', file_client_args=file_client_args),
+    dict(type='LoadAnnotations', with_bbox=True, with_mask=True),
+    dict(type='RandomFlip', prob=0.5),
```
Comment (ZwwWayne) [inline projects/ViTDet/configs/lsj-100e_coco-instance.py:22]: May add comments

Comment (zwhus) [reply projects/ViTDet/configs/lsj-100e_coco-instance.py:22]: it's new config, the new vitdet hasn't been trained yet, I'll use the old version first

File: projects/ViTDet/vitdet/sim_fpn.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+from mmcv.cnn import ConvModule, build_norm_layer
+from mmengine.model import BaseModule
+from torch import Tensor
+
+from mmdet.registry import MODELS
+from mmdet.utils import MultiConfig, OptConfigType
+
+
+@MODELS.register_module()
+class SimFPN(BaseModule):
```
Comment (ZwwWayne) [inline projects/ViTDet/vitdet/sim_fpn.py:15]: rename SimFPN to SimpleFPN

File: projects/ViTDet/REMEAD.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+```bash
+python tools/train.py projects/ViTDet/configs/vitdet_mask-rcnn_vit-b-mae_lsj-100e.py ${CHECKPOINT_PATH}
+```
+
+## Results
+
+Based on mmdetection, this project aligns the test accuracy of the [ViTDet](https://github.com/facebookresearch/detectron2/tree/main/projects/ViTDet).
+<br>
+If you want to reproduce the test results, you need to convert model weights first, then run the test command.
```
Comment (ZwwWayne) [inline projects/ViTDet/REMEAD.md:29]: might need update

Comment (austinmw) [reply projects/ViTDet/REMEAD.md:29]: tiny thing, but noticed README is spelled REMEAD 😝

Comment (zwhus) [reply projects/ViTDet/REMEAD.md:29]: sorry, the new version will update

File: projects/ViTDet/README.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
+# ViTDet
+
+## Description
+
+This is an implementation of [ViTDet](https://github.com/facebookresearch/detectron2/tree/main/projects/ViTDet) based on [MMDetection](https://github.com/open-mmlab/mmdetection/tree/3.x), [MMCV](https://github.com/open-mmlab/mmcv), and [MMEngine](https://github.com/open-mmlab/mmengine).
```
Comment (hhaAndroid) [inline projects/ViTDet/README.md:5]:
```suggestion
This is an implementation of [ViTDet](https://github.com/facebookresearch/detectron2/tree/main/projects/ViTDet) based on [MMDetection](https://github.com/open-mmlab/mmdetection/), [MMCV](https://github.com/open-mmlab/mmcv), and [MMEngine](https://github.com/open-mmlab/mmengine).
```

File: projects/ViTDet/README.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+### Testing commands
+
+In MMDetection's root directory, run the following command to test the model:
+
+```bash
+python tools/test.py projects/ViTDet/configs/vitdet_mask-rcnn_vit-b-mae_lsj-100e.py ${CHECKPOINT_PATH}
+```
+
+If you want to know about how to scaling learning rate and more detailed usage of `test.py/dist_test.sh/slurm_test.sh`, please refer to this [document](../../docs/en/user_guides/test.md)
```
Comment (hhaAndroid) [inline projects/ViTDet/README.md:35]: It should not be necessary

File: projects/ViTDet/vitdet/visiontransformer.py [hunk 7]
```diff
@@ -0,0 +1,448 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
```
Comment (hhaAndroid) [inline projects/ViTDet/vitdet/visiontransformer.py:1]: projects/ViTDet/vitdet/visiontransformer.py -> projects/ViTDet/vitdet/vit.py

File: projects/ViTDet/vitdet/layer_decay_optimizer_constructor.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (hhaAndroid) [inline projects/ViTDet/vitdet/layer_decay_optimizer_constructor.py:1]: Please confirm if you want to rewrite

File: projects/ViTDet/configs/lsj-100e_coco-instance.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        'num_layers': 12,
+    },
+    optimizer=dict(
+        type='AdamW',
+        lr=0.0001,
+        betas=(0.9, 0.999),
+        weight_decay=0.1,
+    ))
+
+max_iters = 184375
```
Comment (hhaAndroid) [inline projects/ViTDet/configs/lsj-100e_coco-instance.py:93]: epoch =？

File: projects/ViTDet/configs/lsj-100e_coco-instance.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+dynamic_intervals = [(max_iters // interval * interval + 1, max_iters)]
+param_scheduler = [
+    dict(
+        type='LinearLR', start_factor=0.001, by_epoch=False, begin=0, end=250),
+    dict(
+        type='MultiStepLR',
+        begin=0,
+        end=max_iters,
+        by_epoch=False,
+        milestones=[163889, 177546],
```
Comment (hhaAndroid) [inline projects/ViTDet/configs/lsj-100e_coco-instance.py:104]: epoch =?

File: projects/ViTDet/configs/lsj-100e_coco-instance.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    max_iters=max_iters,
+    val_interval=interval,
+    dynamic_intervals=dynamic_intervals)
+val_cfg = dict(type='ValLoop')
+test_cfg = dict(type='TestLoop')
+
+default_hooks = dict(
+    logger=dict(type='LoggerHook', interval=50),
+    checkpoint=dict(
+        type='CheckpointHook',
```
Comment (hhaAndroid) [inline projects/ViTDet/configs/lsj-100e_coco-instance.py:119]: You should set max_ckpt, otherwise too many weights are saved

**Comment by Vanish1112** [comment]
The visiontransformer implementation in the vitdet project is different from that in mmpretrain. Can you provide the mmpretrain version of the project? Because, if I directly replace the mae pre-trained visiontransformer in mmpretrain into this project, it is not occupied by shift windows video memory, and more importantly, the Map is far lower than the implementation in this project.

**Comment by cxxgtxy** [comment]
> The visiontransformer implementation in the vitdet project is different from that in mmpretrain. Can you provide the mmpretrain version of the project? Because, if I directly replace the mae pre-trained visiontransformer in mmpretrain into this project, it is not occupied by shift windows video memory, and more importantly, the Map is far lower than the implementation in this project.

VisionLLaMA utilizes both mmpretrain (MAE)and mmdetection (VitDET), which may be helpful for you.

https://github.com/Meituan-AutoML/VisionLLaMA/blob/main/mmdetection/projects/ViTDet/configs/lamadet_mask-rcnn_vit-b-mae_lsj-800-36e.py

**Pull request merged**